File: examples/nodejs/callback.ts
/** Example: Use a callback to log every step of the agent's execution. */

import { Agent, resolveConfig, Event } from '../src/index.js';

async function main() {
  function onEvent(event: Event) {
    if (event.type === 'tool_use_start') {
      console.log(`[Callback] Tool '${event.data.name}' called`);
    } else if (event.type === 'tool_use_end') {
      const result = event.data.result as { isError: boolean };
      const status = result.isError ? 'ERROR' : 'OK';
      console.log(`[Callback] Tool '${event.data.name}' finished: ${status}`);
    } else if (event.type === 'error') {
      console.log(`[Callback] Error: ${event.data.message}`);
    }
  }

  const config = resolveConfig({
    apiKey: process.env.ANTHROPIC_AUTH_TOKEN || '',
    callback: onEvent,
  });

  const agent = new Agent(config);

  for await (const event of agent.run('Read README.md and summarize it')) {
    if (event.type === 'message_delta') {
      process.stdout.write(event.data.text as string);
    } else if (event.type === 'complete') {
      console.log('\n\n[Complete]', event.data.final_content);
    }
  }
}

main().catch(console.error);

File: examples/go/callback.go
// Example: Use a callback to log every step of the agent's execution.
package main

import (
	"context"
	"fmt"
	"os"

	"github.com/steven0lisa/agent-libs/go"
)

func main() {
	cfg := agentlib.DefaultConfig()
	cfg.APIKey = os.Getenv("ANTHROPIC_AUTH_TOKEN")

	// Set a callback to log every event.
	cfg.Callback = func(event agentlib.Event) {
		switch event.Type {
		case agentlib.EventToolUseStart:
			fmt.Printf("[Callback] Tool '%s' called\n", event.Data["name"])
		case agentlib.EventToolUseEnd:
			result := event.Data["result"].(agentlib.ToolResult)
			status := "OK"
			if result.IsError {
				status = "ERROR"
			}
			fmt.Printf("[Callback] Tool '%s' finished: %s\n", event.Data["name"], status)
		case agentlib.EventError:
			fmt.Printf("[Callback] Error: %s\n", event.Data["message"])
		}
	}

	agent := agentlib.NewAgent(cfg)
	ctx := context.Background()

	for event := range agent.Run(ctx, "Read README.md and summarize it") {
		switch event.Type {
		case agentlib.EventMessageDelta:
			fmt.Print(event.Data["text"])
		case agentlib.EventComplete:
			fmt.Printf("\n\n[Complete] %s\n", event.Data["final_content"])
		}
	}
}
